### PR TITLE
Fixed bug 409 due to bad handling of NaN values in zoom linesearch

### PR DESCRIPTION
**Pull request by vroulet**
Fixed bug [409](https://github.com/google/jaxopt/issues/409), added corresponding test, enhanced debugging messages (for linesearch and solvers).

### CODE REVIEW

File: tests/lbfgs_test.py [hunk 1]
```diff
@@ -358,6 +359,56 @@ def _run_lbfgs_solver(X, y, lmbd, n_iter):
...
+
+    y = simulate_binary_logit(x, beta)
+    y = jnp.ravel(y)
+
+    # numpy version
+    def binary_logit_log_likelihood(beta, y,x):
+      lambda_xb = onp.exp(x@beta) / (1 + onp.exp(x@beta))
+      ll_i = y * onp.log(lambda_xb) + (1-y) * onp.log(1-lambda_xb)
+      ll = -onp.sum(ll_i)
+      return ll
```
Comment (mblondel) [inline tests/lbfgs_test.py:391]: remove

File: tests/lbfgs_test.py [hunk 2]
```diff
@@ -358,6 +359,56 @@ def _run_lbfgs_solver(X, y, lmbd, n_iter):
...
+      ll_i = y * jnp.log(lambda_xb) + (1-y) * jnp.log(1-lambda_xb)
+      ll = -jnp.sum(ll_i)
+      return ll
+
+    beta_init = jnp.array([0.01,0.01])
+
+    # using scipy
+    scipy_res = opt.minimize(fun=binary_logit_log_likelihood, 
+                             args=(onp.asarray(y),onp.asarray(x)), 
+                             x0 = (onp.asarray(beta_init)), method='BFGS')
```
Comment (mblondel) [inline tests/lbfgs_test.py:405]: remove

File: jaxopt/_src/zoom_linesearch.py [hunk 3]
```diff
@@ -106,17 +106,17 @@ def _set_val(x, y):
 
 def _check_failure_status(fail_code):
   """Print failure reason according to fail value."""
-  if fail_code == 1:
+  if fail_code % 2 == 0:
```
Comment (mblondel) [inline jaxopt/_src/zoom_linesearch.py:109]: Using bitwise flags would be more standard (I think).

File: tests/lbfgs_test.py (hunk 1 above)
Comment (vroulet) [reply tests/lbfgs_test.py:391]: Which part do you want me to remove exactly? Having the comparison with scipy looks like a good safeguard. By the way I think we can remove the test test_first_stepsize_inf which did not reveal all issues.

Comment (mblondel) [reply tests/lbfgs_test.py:391]: I thought it's sufficient to just check that we don't get stuck at zero (our test suite is getting longer and longer to run). But you're right, maybe it's better to compare with scipy to be entirely sure.

Comment (vroulet) [reply tests/lbfgs_test.py:352]: Ok, I removed the previous test as it was redundant and less documented and kept the comparison to scipy.

**Comment by vroulet** [comment]
I've also modified the way the linesearch would handle the case when the max_stepsize was reached. Before, once the maximal stepsize was reached the algorithm would keep on trying to increase it to find an interval for which the curvature error would be satisfied, while always staying to the max_stepsize. So there would be a waste of function/gradient evaluations. I fixed this by letting the linesearch stop if the max_stepsize is reached and no interval was found (meaning that the sufficient decrease condition is satisfied but not the curvature one and the latter would not be satisfied anyway). I documented it in the code.

File: jaxopt/_src/zoom_linesearch.py [hunk 4]
```diff
@@ -138,7 +150,7 @@ class ZoomLineSearchState(NamedTuple):
 
   error: float
   done: bool
-  fail_code: int  # encode failure status, see _check_status
+  error_flag: int  # encode failure status, see _check_status
```
Comment (mblondel) [inline jaxopt/_src/zoom_linesearch.py:153]: What's the motivation for changing the name? Since, other solvers use `fail_code`, this would create an inconsistency.

Comment (vroulet) [reply jaxopt/_src/zoom_linesearch.py]: The NaN/Inf flag does not necessarily lead to a failure but was more of a warning. One could a priori fully remove it now too. Other than that, we can use fail_code. 
PS: Which other algorithms use fail_code? I could further harmonize if you want.

Comment (mblondel) [reply jaxopt/_src/zoom_linesearch.py]: But in that case error != warning too. `fail_code` is only used in the zoom linesearch but `failed_linesearch` is used in many places so I would prefer to keep `fail_code` for consistency.

Comment (vroulet) [reply jaxopt/_src/zoom_linesearch.py]: Ok! The changes have been made, I've squashed everything.
